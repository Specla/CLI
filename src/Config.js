import path from 'path'
import fs from 'fs'
import dotenv from 'dotenv'
import Autoloader from '@specla/autoloader'
import dot from '@specla/dot-string'

export default class Config {
  /**
   * Create a new configuration
   * @param  {Object} config [description]
   * @return {Config}        [description]
   */
  constructor (config = {}) {
    this._flattenConfig = {}
    this._config = {}
    this._parse(config)
    this._setupDotEnv()
    this._loadConfigFiles()
  }

  /**
   * Load and setup the .env file from the project root
   * @private
   */
  _setupDotEnv () {
    if (!fs.existsSync(path.resolve(process.cwd(), '.env'))) {
      return
    }

    dotenv.config()
  }

  /**
   * Load configfiles from the config path and merge it with the current config
   * @private
   */
  _loadConfigFiles () {
    if (!this.get('config.path')) {
      return
    }

    const configs = new Autoloader(
      path.resolve(this.get('config.path'))
    )

    for (const key in configs) {
      if (typeof configs[key] === 'object' && configs[key].__esModule) {
        configs[key] = configs[key].default
      }
    }

    this._parse({ ...configs, ...this._config })
  }

  /**
   * Parse the config object and create a flatten config
   * @param  {Object} config
   * @return {Object}
   */
  _parse (config) {
    this._flatten(config)

    for (const key in this._flattenConfig) {
      this.set(key, this._flattenConfig[key])
    }
  }

  /**
   * Flatten objects into an object of dot strings
   * @param  {Mixed} config
   * @param  {Array} context
   * @return {Mixed}
   * @private
   */
  _flatten (config, context = []) {
    if (!Array.isArray(config) && typeof config === 'object') {
      return this._flattenObject(config, context)
    }

    return (this._flattenConfig[context.join('.')] = config)
  }

  /**
   * Flatten object
   * @param  {Mixed} config
   * @param  {Array} context
   * @private
   */
  _flattenObject (config, context) {
    for (const key in config) {
      this._flatten(config[key], context.concat(key))
    }
  }

  /**
   * Get configuration from dot string
   * @param  {String} key
   * @return {Mixed}
   * @public
   */
  get (key) {
    try {
      return dot.get(this._config, key)
    } catch (err) {
      return undefined
    }
  }

  /**
   * Set value for a given key
   * @param {String} key
   * @param {Mixed} value
   * @return {Mixed}
   * @public
   */
  set (key, value) {
    if (!this._flattenConfig[key]) {
      this._flattenConfig[key] = value
    }

    return dot.set(this._config, key, value)
  }
}
