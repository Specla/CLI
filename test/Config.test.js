/* eslint-env jest */
import fs from 'fs'
import { exec } from 'child_process'
import Config from '../src/Config'
import setup from './utils/setup'
import cleanup from './utils/cleanup'

beforeEach(setup)
afterEach(cleanup)

test('Should load the .env file if it exists in the project root', () => {
  new Config() // eslint-disable-line
  expect(process.env.APP_HOST).toBe('localhost')
  delete process.env.APP_HOST
})

test('Should work without the .env file', () => {
  fs.unlinkSync('.env')
  new Config() // eslint-disable-line
  expect(process.env.APP_HOST).toBe(undefined)
  delete process.env.APP_HOST
})

test('Should load config files from the config folder and merge them together', () => {
  const config = new Config({ 'config.path': './config' })
  expect(config).toMatchSnapshot()
})

test('Should merge the config argument in the constructor with the config files', () => {
  const config = new Config({ app: { host: '127.0.0.1' } })
  expect(config.get('app.host')).toBe('127.0.0.1')
})

test('Should access item in the config object', () => {
  const config = new Config({ mySettings: true })
  expect(config.get('mySettings')).toBe(true)
})

test('Should set new item in the config object', () => {
  const config = new Config()
  config.set('app.setting', true)
  expect(config._flattenConfig['app.setting']).toBe(true)
  expect(config._config.app.setting).toBe(true)
})

test('Should work without a config folder', done => {
  exec('rm -rf ./config', () => {
    const config = new Config({ setting: true })
    expect(config).toMatchSnapshot()
    done()
  })
})
