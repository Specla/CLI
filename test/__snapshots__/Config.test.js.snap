// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should load config files from the config folder and merge them together 1`] = `
Config {
  "_config": Object {
    "app": Object {
      "array": Array [
        "item1",
        "item2",
      ],
      "default": "test",
      "host": "localhost",
      "locale": Object {
        "dateFormat": "YY-mm-dd",
        "language": "en-US",
      },
    },
    "config": Object {
      "path": "./config",
    },
  },
  "_flattenConfig": Object {
    "app.array": Array [
      "item1",
      "item2",
    ],
    "app.default": "test",
    "app.host": "localhost",
    "app.locale.dateFormat": "YY-mm-dd",
    "app.locale.language": "en-US",
    "config.path": "./config",
  },
}
`;

exports[`Should work without a config folder 1`] = `
Config {
  "_config": Object {
    "setting": true,
  },
  "_flattenConfig": Object {
    "setting": true,
  },
}
`;
